flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/active/ActiveResourceManagerTest.java

```diff
@@ -0,0 +1,428 @@
+					// finishes before the assertions
+					return null;
+				}).get(TIMEOUT_SEC, TimeUnit.SECONDS);
+				assertFalse(requestWorkerFromDriverFutures.get(1).isDone());
+			});
+		}};
+	}
+
+	/**
+	 * Tests workers from previous attempt successfully recovered and registered.
+	 */
+	@Test
+	public void testRecoverWorkerFromPreviousAttempt() throws Exception {
+		new Context() {{
+			final ResourceID tmResourceId = ResourceID.generate();
+
+			runTest(() -> {
+				runInMainThread(() -> getResourceManager().onPreviousAttemptWorkersRecovered(Collections.singleton(tmResourceId)));
+				CompletableFuture<RegistrationResponse> registerTaskExecutorFuture = registerTaskExecutor(tmResourceId);
+				assertThat(registerTaskExecutorFuture.get(TIMEOUT_SEC, TimeUnit.SECONDS), instanceOf(RegistrationResponse.Success.class));
+			});
+		}};
+	}
+
+	/**
+	 * Tests decline unknown worker registration.
+	 */
+	@Test
+	public void testRegisterUnknownWorker() throws Exception {
+		new Context() {{
+			runTest(() -> {
+				CompletableFuture<RegistrationResponse> registerTaskExecutorFuture = registerTaskExecutor(ResourceID.generate());
+				assertThat(registerTaskExecutorFuture.get(TIMEOUT_SEC, TimeUnit.SECONDS), instanceOf(RegistrationResponse.Decline.class));
+			});
+		}};
+	}
+
+	@Test
+	public void testOnError() throws Exception {
+		new Context() {{
+			final Throwable fatalError = new Throwable("Testing fatal error");
+			runTest(() -> {
+				runInMainThread(() -> getResourceManager().onError(fatalError));
+				final Throwable reportedError = getFatalErrorHandler().getErrorFuture().get(TIMEOUT_SEC, TimeUnit.SECONDS);
+				assertThat(reportedError, is(fatalError));
+			});
+		}};
+	}
+
+	class Context {
+
+		final Configuration flinkConfig = new Configuration();
+		final TestingResourceManagerDriver.Builder driverBuilder = new TestingResourceManagerDriver.Builder();
+		final TestingSlotManagerBuilder slotManagerBuilder = new TestingSlotManagerBuilder();
+
+		private ActiveResourceManager<ResourceID> resourceManager;
+		private TestingFatalErrorHandler fatalErrorHandler;
+
+		ActiveResourceManager<ResourceID> getResourceManager() {
+			return resourceManager;
+		}
+
+		TestingFatalErrorHandler getFatalErrorHandler() {
+			return fatalErrorHandler;
+		}
+
+		void runTest(RunnableWithException testMethod) throws Exception {
+			fatalErrorHandler = new TestingFatalErrorHandler();
+			resourceManager = createAndStartResourceManager(
+					flinkConfig,
+					driverBuilder.build(),
+					slotManagerBuilder.createSlotManager());
+
+			try {
+				testMethod.run();
+			} finally {
+				resourceManager.close();
+			}
+		}
+
+		private ActiveResourceManager<ResourceID> createAndStartResourceManager(
+				Configuration configuration,
+				ResourceManagerDriver<ResourceID> driver,
+				SlotManager slotManager) throws Exception {
+			final TestingRpcService rpcService = new TestingRpcService(configuration);
```

Comment:
nit: I think it would be a bit cheaper to reuse the `TestingRpcService` via `@ClassRule static final TestingRpcServiceResource testingRpcServiceResource = ....`.

Reply:
The `configuration` used for creating `TestingRpcService` might be different for test cases.
It's probably not a problem, since the chance that `configuration` contains something affects the `TestingRpcService` is small. But I would still like to make it save.

Reply:
Hmm, do any of the tests configure something `RpcService` related? If a new test should require this, then one would probably create a separate test class since all other tests can work with the same `RpcService`.

Reply:
Ok, make sense. I could also add a notice about this.